crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -1080,74 +1144,537 @@ fn format_docstring(normalized: &NormalizedString, f: &mut PyFormatter) -> Forma
+fn doctest_find_ps2_prompt<'src>(ps1_indent: &str, line: &'src str) -> Option<&'src str> {
+    let (ps2_indent, ps2_after) = line.split_once("...")?;
+    // PS2 prompts must have the same indentation as their
+    // corresponding PS1 prompt.[1] While the 'doctest' Python
+    // module will error in this case, we just treat this line as a
+    // non-doctest line.
+    //
+    // [1]: https://github.com/python/cpython/blob/0ff6368519ed7542ad8b443de01108690102420a/Lib/doctest.py#L733
+    if ps1_indent != ps2_indent {
+        return None;
+    }
+    // PS2 prompts must be followed by an ASCII space character unless
+    // it's an otherwise empty line[1].
+    //
+    // [1]: https://github.com/python/cpython/blob/0ff6368519ed7542ad8b443de01108690102420a/Lib/doctest.py#L809-L812
+    match ps2_after.strip_prefix(' ') {
+        None if ps2_after.is_empty() => Some(""),
+        None => None,
+        Some(code) => Some(code),
+    }
+}
+
+/// Formats the given source code using the given options.
+///
+/// The given quote style should correspond to the style used by the docstring
+/// containing the code snippet being formatted. The formatter will use this
+/// information to invert the quote style of any such strings contained within
+/// the code snippet in order to avoid writing invalid Python code.
+///
+/// This is similar to the top-level formatting entrypoint, except this
+/// explicitly sets the context to indicate that formatting is taking place
+/// inside of a docstring.
+fn docstring_format_source(
+    options: &crate::PyFormatOptions,
+    docstring_quote_style: QuoteStyle,
+    source: &str,
+) -> Result<Printed, FormatModuleError> {
+    use ruff_python_parser::AsMode;
+
+    let source_type = options.source_type();
+    let (tokens, comment_ranges) = ruff_python_index::tokens_and_ranges(source, source_type)?;
+    let module =
+        ruff_python_parser::parse_ok_tokens(tokens, source, source_type.as_mode(), "<filename>")?;
+    let source_code = ruff_formatter::SourceCode::new(source);
+    let comments = crate::Comments::from_ast(&module, source_code, &comment_ranges);
+    let locator = Locator::new(source);
+
+    let ctx = PyFormatContext::new(options.clone(), locator.contents(), comments)
+        .in_docstring(docstring_quote_style);
+    let formatted = crate::format!(ctx, [module.format()])?;
+    formatted
+        .context()
+        .comments()
+        .assert_all_formatted(source_code);
+    Ok(formatted.print()?)
+}
+
+/// If the last line of the docstring is `content" """` or `content\ """`, we need a chaperone space
+/// that avoids `content""""` and `content\"""`. This does only applies to un-escaped backslashes,
+/// so `content\\ """` doesn't need a space while `content\\\ """` does.
+fn needs_chaperone_space(normalized: &NormalizedString, trim_end: &str) -> bool {
+    trim_end.ends_with(normalized.quotes.style.as_char())
+        || trim_end.chars().rev().take_while(|c| *c == '\\').count() % 2 == 1
 }
```

Comment:
Slightly superficial comment, but what's your reaction to splitting the docstring-specific code out into its own file?

Reply:
I thought about this while coding and I was torn. I think I ultimately didn't do it because I thought that we might eventually want to refactor it further (possibly) so that it can be used in other contexts aside from formatting Python code. So I didn't want to spend too much time on making it a "proper" abstraction. With that said, I do believe it lifts pretty cleanly out of this module. I'll try it out and see how it feels.

Reply:
It would also be fine as a separate change / PR after-the-fact if easier and less distracting.